doc/aaa/TACACS+ Requirement.md

```diff
@@ -0,0 +1,161 @@
+
+# 1 Functional Requirement
+## 1.1 Authentication
+- Authentication when user login to SONiC host.
+
+
+## 1.2 Authorization
+- Authorization when:
+	- User login to SONiC host.
+	- User run command on SONiC host.
+		- User can only run commands in whitelist.
+		- For recursive command/script, only the top level command have authorization.
+
+- Only command in whitelist  visible to user:
+	- All privilege level share same whitelist to simplify design.
+	- Different privilege level have different permission to run these command.
+	- All commands in sudoers will add to the whitelist. and sudoers config file still need for RO users, this is because when remote TACACS server not avaliable, we need use local group permission for failover.
+
+- Disable user behavior in shell:
+	- Changing directories with the cd builtin.
+		- User still can access files in other folder, and use 'ls' command to list content in other folder, for example:
+        ```
+            test@testhost:~$ ls -l /etc/
+            total 880
+            drwxr-xr-x 3 test test       4096 Aug 22  2020 NetworkManager
+            drwxr-xr-x 7 test test       4096 Aug 22  2020 X11
+            drwxr-xr-x 3 test test       4096 Aug 22  2020 acpi
+        ```
+	- Setting or unsetting the values of the SHELL, PATH, HISTFILE, ENV, or BASH_ENV variables.
+	- Specifying command names containing slashes.
+	- Importing function definitions from the shell environment at startup.
+	- Parsing the value of SHELLOPTS from the shell environment at startup.
+	- Redirecting output using the ‘>’, ‘>|’, ‘<>’, ‘>&’, ‘&>’, and ‘>>’ redirection operators.
+	- Builtin commands:
+
+| **Command** | **Behavior**                |
+| -------- | -------------------------- |
+| . | Specifying a filename containing a slash as an argument.     |
+| history     | Specifying a filename containing a slash as an argument. |
+| hash     | Specifying a filename containing a slash as an argument to the -p option. |
+| exec     | Specifying a filename containing a slash as an argument to the -p option. |
+| deleting/adding     | Use the -f and -d options to the enable builtin. |
+| enable     | Using the 'enable' builtin command to enable disabled shell builtins. |
+| command     | Specifying the -p option to the 'command' builtin command. |
+| set     | Turning off restricted mode with ‘set +r’ or ‘set +o restricted’. |
+
+	- All these behavior disabled only for user input, command in script will not be affected, see here for more details: https://www.gnu.org/software/bash/manual/html_node/The-Restricted-Shell.html
+
+
+- Supported Authorization types:
+	- EXEC: user session authorization support. this happen when user login.
+	- Command: user run command in shell.
+
+- Failover:
+	- Authorization will happen before execute, if remote TACACS server not available, use local group based authorization as failover.
+
+
+## 1.3 Accounting
```

Comment:
>Accounting [](http://example.com/codeflow?start=7&length=10)

Could you explain what is accounting and what will be accounted?

Reply:
Will update document, accounting basically record what user doing, and following event will accounted:
user login/logout, user start run a command, and user command finish.